L3 Licensed Miners Database

PJLewandowski opened the pull request.

PJLewandowski committed:
Implement structure to hold miners license

PJLewandowski committed:
Handle new license transaction

PJLewandowski committed:
Dump/load miner's licenses database

PJLewandowski committed:
Unit tests

PJLewandowski committed:
Little cleaning before opening PR

mtwaro commented on src/test/ddms_tests.cpp:
```diff
@@ -0,0 +1,233 @@
+	tx.vout.resize(2);
+	tx.vout[0].scriptPubKey = CScript();
+	tx.vout[0].nValue = 49000;
+	tx.vout[1].scriptPubKey = CScript() << OP_RETURN << std::vector<unsigned char>{3, 76, 84, 120, 20, 169, 20, 230, 53, 247, 106, 15, 189, 177, 112, 86, 88, 161, 227, 176, 86, 115, 120, 7, 74, 0, 5};
```
Might be worthy to split the vector into logical parts (i.e. header, script_size, script, hashrate) and extract them into constants. Also, please specify the values in hex format.
One more thing - are you sure that the 5th value in the vector (20) is correct? It should specify the script length + 2, but the script itself is 23 bytes long.

PJLewandowski replied:
Done (structure of LTx CScript changed)

mtwaro commented on src/test/ddms_tests.cpp:
```diff
@@ -0,0 +1,233 @@
+	g_txindex = MakeUnique<TxIndex>(1 << 20, true);
+	g_txindex->Start();
+
+	CScript fakeWdmoScript = WDMO_SCRIPT; --fakeWdmoScript[1];
```
`--fakeWdmoScript[1]` will cause failure in the script syntax itself, since this byte specifies the length of script hash pushed to the stack. To keep it cleaner, modify the values in the hash itself, e.g. byte no 10.

PJLewandowski replied:
Done

mtwaro commented on src/validation.cpp:
```diff
@@ -225,6 +224,8 @@ class CChainState {
  */
 RecursiveMutex cs_main;
 
+const CScript WDMO_SCRIPT = CScript() << OP_HASH160 << std::vector<unsigned char>{11, 182, 127, 3, 232, 176, 211, 69, 45, 165, 222, 55, 211, 47, 198, 174, 240, 165, 160, 160} << OP_EQUAL; // TODO: replace with actual wdmo script; use ParseHex
```
`validation.cpp` is already too big, we should extend it as little as possible. This definition, along with MinerLicences should be moved to some specific directory, e.g. `src/policy/ddms`

PJLewandowski replied:
Done

mtwaro commented on src/validation.cpp:
```diff
@@ -322,6 +323,84 @@ static void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfte
+	if (!license || !NeedToUpdateLicense(entry))
+		return;
+
+	license->hashRate = entry.hashRate;
```
I would also add a case for `hashRate==0`, which would result in removing the license from database

PJLewandowski replied:
Done (with new unit test)

PJLewandowski committed:
PR fixes:

- moving MinerLicenses to policy/ddms
- removing license when assigning 0 hashrate

PJLewandowski committed:
PR fixes:

- changing structure of LTx CScript
